### ISSUE TITLE
strace: __NR_newfstatat undeclared

### ISSUE THREAD

**Issue by frikiluser** [landley/toybox#368]
Hi,

After #298...

While it builds fine on amd64, trying to build strace toy under i386 Debian Sid GCC-12 throws the following error:
```
toys/pending/strace.c: In function 'print_enter':
toys/pending/strace.c:471:22: error: '__NR_newfstatat' undeclared (first use in this function)
  471 | #define SC(n,f) case __NR_ ## n: name = #n; TT.fmt = f; break
      |                      ^~~~~
toys/pending/strace.c:500:5: note: in expansion of macro 'SC'
  500 |     SC(newfstatat, "fs/{stat}d");
      |     ^~
toys/pending/strace.c:471:22: note: each undeclared identifier is reported only once for each function it appears in
  471 | #define SC(n,f) case __NR_ ## n: name = #n; TT.fmt = f; break
      |                      ^~~~~
toys/pending/strace.c:500:5: note: in expansion of macro 'SC'
  500 |     SC(newfstatat, "fs/{stat}d");
      |     ^~
```
Config: https://salsa.debian.org/debian/toybox/-/raw/ceef75bd6d2d6aaec1419009a235139a2f4e693b/debian/toybox-almostallyesconfig
Build log: https://salsa.debian.org/debian/toybox/-/jobs/3121734

**Comment by landley** [comment]
Pending again. This command is brittle, incomplete, and only really builds on x86 and arm. I hope to spend a couple weeks on it later this year, but it's way down the todo list...

**Comment by enh-google** [comment]
> only really builds on x86 and arm

actually, it's the other way round... it only really builds on LP64. i think i have a "fixed" version on my raspberry pi, but dealing with LP32's legacy syscalls is inherently quite gross (more #ifdefs) so i've been hoping 32-bit is dead to us by the time it gets out of pending. (certainly i think mixed 32/64 systems should be dead soon -- they're basically dead to me now as SoCs start losing the ability to even run arm32 code, and it looks like the distros are starting to give up on x86 too -- but it seems like we'll have a few more years of 32-only and 64-only coexisting, sadly. not having to support mixed systems at least means we won't have to worry about supporting `strace -f` on a 64-bit process that spawns a 32-bit process or vice versa.)

iirc what stopped me from even sending you the patch was that i couldn't test my arm32 changes on x86 because my old copy & paste for installing the 32-bit sysroot on ubuntu didn't work any more?

**Comment by landley** [comment]
I applied your patch, commit d627fa05d480.

In theory sizeof(long) should match sizeof(pointer) on 32 bit too, but gcc is full of "long becomes int in 32 bit headers so it warns even though the size is the same there". Sigh, legacy scar tissue. I notice most often with sizeof().

Dealing with the "32 bit on 64 bit" compatibility syscalls does indeed seem horrific. Either we need to use peek() and poke() or just punt and say "run a 32 bit build of strace for tracing programs in 32 bit mode". The strace in my tree has several changes but I can redo that from scratch later if necessary, strace not in scope for next release.  (Probably November 12-ish. I'm tempted to track kernel releases for the mkroot builds, but they're doing 9 weeks these days and 3 months is 13 weeks. Not a small difference.)

**Comment by enh-google** [comment]
> Dealing with the "32 bit on 64 bit" compatibility syscalls does indeed seem horrific. Either we need to use peek() and poke() or just punt and say "run a 32 bit build of strace for tracing programs in 32 bit mode". 

i've said it before, but i strongly recommend "punt" ... the real kicker with trying to do better is that then we need to teach toybox about the exact struct layout of every struct for every architecture. this is one of the main reasons why "real" strace is so enormous.

> The strace in my tree has several changes but I can redo that from scratch later if necessary, strace not in scope for next release. (Probably November 12-ish. I'm tempted to track kernel releases for the mkroot builds, but they're doing 9 weeks these days and 3 months is 13 weeks. Not a small difference.)

yeah, let me know whether you want x86 and arm32 support... i'm never sure quite how serious you are about stuff like m68k. (i'm never sure quite how serious debian is about all of amd64/arm64/armel/armhf/i386/mips64el/mipsel/ppc64el/s390x! obviously amd64 is real, and raspberry pi made armhf "somewhat real", even if they're moving to arm64. i'd be curious to see download numbers per architecture.) but even if they're mostly in my rear-view mirror at this point, i've used both x86 and arm32 in the last decade, so i don't consider them comically obsolete in the same way that, say, sparc or powerpc would be (let alone m68k --- it's been a very long time since i last saw a sun2/50!).

**Comment by paulwratt** [comment]
> certainly i think mixed 32/64 systems should be dead soon

I am glad the RaspberryPi Foundation does not feel the same as you ..

.. personally I prefer the unpresidented reliability that a 64bit
Kernel and 32bit user space brings. (see also X32 binaries)

**Comment by enh-google** [comment]
> I am glad the RaspberryPi Foundation does not feel the same as you .. .. personally I prefer the unpresidented reliability that a 64bit Kernel and 32bit user space brings. (see also X32 binaries)

that's irrelevant --- the question is whether you have 32-bit processes exec'ing 64-bit processes or vice versa. that can happen on most current Android devices which have both 32- and 64-bit userspaces, but not on Android Auto which is 64-bit only, nor Android Go which is 32-bit only. for the purposes of this bug, your Raspbian system is 32-only.

given that new Arm 64-bit cores don't support running 32-bit code, the value of shipping the 32-bit userspace drops to zero, so although 32/64 is currently the most common option, everyone's likely to have to "pick a side" in the next few years and be either 32-only (because they're still using armv7 or whatever) or 64-only (because they're using an armv9 core that can't run 32-bit code anyway).
